windows-containers/windows-logging.md

```diff
@@ -0,0 +1,166 @@
+---
+title: OpenShift Windows Worker Node and Container Logging
+authors:
+  - "@ravisantoshgudimetla"
+  - "@aravindhp"
+reviewers:
+  - "@crawford"
+  - "@sdodson"
+  - "@jcantrill"
+  - "@richm" 
+approvers:
+  - "@crawford"
+  - "@sdodson"
+  - "@jcantrill"
+  - "@richm"  
+creation-date: 2019-09-10
+last-updated: 2019-09-10
+status: implementable
+---
+
+# logging-for-windows-worker-nodes
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+The intent of this enhancement is to enable logging for Windows worker nodes 
+and pods running in an OpenShift cluster
+
+## Motivation
+
+Logging is a critical component for identifying issues with nodes, containers
+running on the nodes. The main motivation behind this enhancement is to bring 
+Windows node logging infrastructure on par with linux nodes
+
+### Goals
+
+As part of this enhancement, we plan to do the following:
+* Deploy log collection infrastructure onto Windows nodes
+* Configure container runtime to use the logging infrastructure we have deployed
+* Collect node and pod logs
+* Upgrade log collection infrastructure
+* Forward logs and events from nodes and pods to log store
+
+### Non-Goals
+
+As part of this enhancement, we do not plan to support:
+* Logging for Windows containers
+  * Microsoft will provide us with the logging framework that enables Windows 
+    containers to redirect logs to standard out from various sources like 
+    ETW, Perf Counters etc
+* Customization of log collection infrastructure
+  * We plan to support fluentd only
+  * Microsoft will provide us with a Windows fluentd binary
+
+
+## Proposal
+
+The main idea here is to deploy fluentd as a log collection agent onto Windows 
+nodes after cluster logging operator is deployed onto the cluster. Fluentd has
+support for Windows logging. We plan to leverage it and run fluentd as a service
+running on Windows node.
+
+
+### Implementation Details
+
+We plan to create a Windows Node logging Ansible Playbook that can
+
+* Download fluentd binary for Windows from a known location and transfers it to
+  the Windows node
+* Configure the container runtime to use fluentd as the logging driver
+* Configure fluentd to communicate with elastic search data store in a format
+  similar to logs emitted from linux worker nodes
+
+The inputs to this playbook
+
+* The ip address, port, certificate to connect the elastic search service
+* URL from which fluentd Windows binary can be downloaded from
+
+We can get the ipaddress and other details from various pods and secrets present
+in the OpenShift cluster, once the cluster logging operator has been deployed.
```

Comment:
How will the windows fluentd pod get the kubernetes metadata?  When using docker with the json-file log driver, or when using cri-o, logs are written to the node in the following format: `/var/log/containers/$pod-name_$namespace-name_$container-name-$containerid.log`.  For docker with the journald log driver, similar metadata is provided in the `CONTAINER_NAME` and `CONTAINER_ID_FULL` fields.  Our logging pipeline relies on having access to this metadata.  The logging pipeline must use this to look up the namespace UUID from the Kubernetes API (via the fluent-plugin-kubernetes_metadata_filter), which it uses to construct the Elasticsearch index name `project.$namespace_name.$namespace_uuid.$date`, which must be in this format for multi-tenancy/access control to work.

Reply:
learned fluent is being deployed as a first class service on the node outside of the kublet